packages/beacon-node/src/api/impl/validator/index.ts

```diff
@@ -649,223 +600,326 @@ export function getValidatorApi(
+            chain.assembleBlockBody(BlockType.Blinded, blockAttributes, currentState, commonBlockBody, b.value),
+          ]);
+
+          const result = selectBlockProductionSource({
+            engineBlockValue: e.value.executionPayloadValue + engineBlockResp.consensusBlockValue,
+            builderBlockValue: b.value.executionPayloadValue + builderBlockResp.consensusBlockValue,
+            builderBoostFactor,
+            builderSelection,
+          });
+          const executionPayloadSource = result.source;
 
-      return {...engine.value, executionPayloadBlinded: false, executionPayloadSource: ProducedBlockSource.engine};
-    }
+          if (executionPayloadSource === ProducedBlockSource.engine) {
+            return selectEngineBlock(result.reason, {executionPayloadBlinded: false});
+          }
 
-    if (engine.status === "fulfilled" && builder.status === "fulfilled") {
-      const result = selectBlockProductionSource({
-        builderBlockValue: builder.value.executionPayloadValue + builder.value.consensusBlockValue,
-        engineBlockValue: engine.value.executionPayloadValue + engine.value.consensusBlockValue,
-        builderBoostFactor,
-        builderSelection,
-      });
-      const executionPayloadSource = result.source;
+          return selectBuilderBlock(result.reason, {executionPayloadBlinded: true});
+        },
+      },
+    };
 
-      metrics?.blockProductionSelectionResults.inc(result);
+    return executeMatrix(engineVsBuilderMatrix, engine, builder);
+  }
 
-      logger.info(`Selected ${executionPayloadSource} block`, {
-        reason: result.reason,
-        ...loggerContext,
-        engineDurationMs: engine.durationMs,
-        ...getBlockValueLogInfo(engine.value, ProducedBlockSource.engine),
-        builderDurationMs: builder.durationMs,
-        ...getBlockValueLogInfo(builder.value, ProducedBlockSource.builder),
-      });
+  async function produceEngineOrBuilderBlock({
+    feeRecipient,
+    builderSelection,
+    strictFeeRecipientCheck,
+    slot,
+    randaoReveal,
+    graffiti,
+    builderBoostFactor,
+    forceBlockType,
+  }: routes.validator.ExtraProduceBlockOpts & {
+    slot: Slot;
+    randaoReveal: BLSSignature;
+    graffiti?: string;
+    // TODO deneb: skip randao verification
+    skipRandaoVerification?: boolean;
+    builderBoostFactor?: bigint;
+    forceBlockType?: BlockType;
+  }): Promise<ProduceFullOrBlindedBlockOrContentsRes> {
+    notWhileSyncing();
+    await waitForSlot(slot); // Must never request for a future slot > currentSlot
 
-      if (executionPayloadSource === ProducedBlockSource.engine) {
-        return {
-          ...engine.value,
-          executionPayloadBlinded: false,
-          executionPayloadSource,
-        };
+    const parentBlockRoot = fromHex(chain.getProposerHead(slot).blockRoot);
+    notOnOutOfRangeData(parentBlockRoot);
+    const ctx = normalizeBlockProductionOptions({slot, builderBoostFactor, builderSelection, forceBlockType});
```

Comment:
need better name than this, it's more than just context, it's `BlockProductionOptions`